### PR TITLE
Adds new Design System to Text Input

### PR DESCRIPTION
**Pull request by martins87**
## Link to Issue
Closes: #3935 

## Description of Changes
- Adds new design system to TextInput component

## Test Plan
- CA (click around) tested on local and frack:
- New component was tested on Storybook and on some application pages

### CODE REVIEW

**Comment by martins87** [comment]
@zakhap @meeshlin please review it at your earliest convenience

**Comment by martins87** [comment]
Moving to Draft after Design Review

**Review by zakhap — COMMENTED** [review]
This looks good to me upon an inspection review! 

Anything you want to critique on this PR @meeshlin ?

**Comment by meeshlin** [comment]
This needs further review, will need the day to do this as it's mostly around behavior. @martins87 @zakhap

**Comment by meeshlin** [comment]
**Design Feedback**
- The container with only the text has a line-height and regular height of 20px, however, with the padding around it, the greater container has a height of **32px for `isCompact=Yes` (`padding: 6px 8px;`)** and 40px for `isCompact=No` (`padding: 10px 12px;`). 
- Typography should be the same for both, which is `/* Body 02/Regular */` (`font-family: Neue Haas Unica; font-size: 14px; font-style: normal; font-weight: 400; line-height: 20px; letter-spacing: 0.14px;`)

@martins87 @zakhap

**Comment by meeshlin** [comment]
It seems like the actual height rather than the font's line-height (20px) is built into the text component—making them 32px and 40px with the added padding, which drastically changes the size. The actual text input for both is the same, it's just the padding that is different. 

**From Chromatic:**
Screenshot 2023-07-07 at 2 49 36 PM

Screenshot 2023-07-07 at 2 49 56 PM

@martins87

**Comment by martins87** [comment]
> **Design Feedback**
> 
> * The container with only the text has a line-height and regular height of 20px, however, with the padding around it, the greater container has a height of **32px for `isCompact=Yes`** (`padding: 6px 8px;`) and **40px for `isCompact=No`** (`padding: 10px 12px;`).
> * Typography should be the same for both, which is **`/* Body 02/Regular */`** (`font-family: Neue Haas Unica; font-size: 14px; font-style: normal; font-weight: 400; line-height: 20px; letter-spacing: 0.14px;`)
> 
> @martins87 @zakhap

Addressed.

**Comment by martins87** [comment]
> It seems like the actual height rather than the font's line-height (20px) is built into the text component—making them 32px and 40px with the added padding, which drastically changes the size. The actual text input for both is designed to be the same (**`/* Body 02/Regular */`**), it's just the padding that is different.
> 
> **From Chromatic:** Screenshot 2023-07-07 at 2 49 36 PM
> 
> Screenshot 2023-07-07 at 2 49 56 PM
> @martins87

Addressed.

**Comment by meeshlin** [comment]
Checked Chromatic and the only thing I see that's incorrect is the `Large` text input. It should have the same font size in the component as the small (see [here](https://www.figma.com/file/eIVp33a1oCu0AtcLwSbGjr/%F0%9F%9A%A7-Components-and-Patterns?type=design&node-id=339%3A5159&mode=dev)).

@martins87

**Comment by martins87** [comment]
> Checked Chromatic and the only thing I see that's incorrect is the `Large` text input. It should have the same font size in the component as the small (see [here](https://www.figma.com/file/eIVp33a1oCu0AtcLwSbGjr/%F0%9F%9A%A7-Components-and-Patterns?type=design&node-id=339%3A5159&mode=dev)).
> 
> @martins87

Addressed.

**Comment by masvelio** [comment]
- right icon is buggy
https://github.com/hicommonwealth/commonwealth/assets/14819225/0a2295d1-68a8-4af8-ad14-a84682c22c61

**Comment by martins87** [comment]
> * right icon is buggy
> 
>  Kapture.2023-07-18.at.12.21.11.mp4

Addressed.

File: packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss
Comment (masvelio) [inline packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss:37]: by doing this, you are overriding existing Input styles. It is intended behaviour? Do we want to just create new input in this PR or we want to Create and Replace it across the board? Right know it is messed up because Input from new_designs is used only in component showcase, however, the styles are overriding the old input what breaks the app in many places. cc @zakhap 
image

image

image

**Comment by masvelio** [comment]
> > * right icon is buggy
> > 
> > Kapture.2023-07-18.at.12.21.11.mp4
> 
> Addressed.

image

does not seem to be?

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_text_input.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_text_input.tsx:82]: I am looking on the figma and validation text is visible under the input. 
image

In your solution validation text is above 
image

Also some icons are missing? And the informational text?

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_text_input.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_text_input.tsx:149]: as I mentioned couple times already, do not inject inline style => move this code to css

**Comment by CowMuon** [comment]
Looks like we still have unresolved comments before we can merge this in? @martins87

**Comment by martins87** [comment]
> Looks like we still have unresolved comments before we can merge this in? @martins87 

Yes. I'm working on them

File: packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss
Comment (masvelio) [inline packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss:43]: we do not need comments like these. It is obvious what this property does

**Comment by martins87** [comment]
> > > * right icon is buggy
> > > 
> > > Kapture.2023-07-18.at.12.21.11.mp4
> > 
> > 
> > Addressed.
> 
> image
> 
> does not seem to be? I am looking on [chromatic](https://645a5781c0021a5d05d60c23-phkxjhbibj.chromatic.com/?path=/story/components-form-fields-textinput--small&args=iconLeft:airplane;iconLeftColor:purple-200;iconRight:alarm;iconRightColor:green-400)

Addressed: https://645a5781c0021a5d05d60c23-thcitrthwx.chromatic.com/?path=/story/components-form-fields-textinput--small&args=iconRight:alignCenterVertical

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_text_input.tsx
Comment (martins87) [reply packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_text_input.tsx]: Addressed.

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_text_input.tsx
Comment (martins87) [reply packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_text_input.tsx]: Addressed validation message, icon and instructional message.

File: packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss
Comment (martins87) [reply packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss]: > 

I changed the old 'TextInput' class to 'OldTextInput', since it will probably not going to be used in the future.

@zakhap are we changing all TextInputs to the new design system one?

File: packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss
Comment (martins87) [reply packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss]: Addressed.

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_text_input.tsx
Comment (martins87) [reply packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_text_input.tsx]: I tried some solutions but apparently there is no way to edit a css class property without editing it inline. It's a easy thing to by editing it inline on the React file as the previous solution... or use some libraries like [styled-components](https://styled-components.com/).

One solution I thought was setting the default width as 240px (as per design) and have the `fullWidth` prop so the width can be set in the outer div. Example:

image

image

@zakhap @meeshlin @masvelio what do you guys think?

Comment (masvelio) [reply packages/commonwealth/client/scripts/views/components/component_kit/new_designs/cw_text_input.tsx]: yeah and this is the right solution. Default is 240 but if you want to check the width of the input, you simply wrap it with the div and pass the `fullWidth` prop into the input.

File: packages/commonwealth/client/styles/components/component_kit/new_designs/CWTextInput.scss
Comment (masvelio) [inline packages/commonwealth/client/styles/components/component_kit/new_designs/CWTextInput.scss:62]: not sure why we introduce variables here? in this case `input` will always be `width: 100%`. `input-and-icon-container` will be 240 by default and might be overridden with `fullWidth` class.

File: packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss
Comment (masvelio) [reply packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss]: Apparently changing it to OldTextInput is not enough. Some styles are still messed up here and there, because your inputs are not 100% in isolation. 

1. edit profile page: 
- vertical margin are messed up
- there is error but green text on validation text
- validation text is misaligned

image

2. create thread page:
- both inputs are full width in block styles

image

In both cases your changes are affecting the old styles which is not desired. Probably there are more places like this in the app where inputs are messed up

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTextInput.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTextInput.tsx:104]: do we need the information if user is typing? Figma indeed have typing as different state but it looks exactly the same as focus. The same as default vs filled - they are the same https://www.figma.com/file/eIVp33a1oCu0AtcLwSbGjr/%F0%9F%9A%A7-Components-and-Patterns?type=design&node-id=493%3A4751&mode=dev

**Review by masvelio — COMMENTED** [review]
Component itself looks pretty ready but new input affects the old input styles so there are places in the app that are broken in styles which I guess is not desirable

File: packages/commonwealth/client/styles/components/component_kit/new_designs/CWTextInput.scss
Comment (martins87) [reply packages/commonwealth/client/styles/components/component_kit/new_designs/CWTextInput.scss]: [Removed](https://github.com/hicommonwealth/commonwealth/commit/42cdff4965a362fc768f44280d56af5dc99f0027).

File: packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss
Comment (martins87) [reply packages/commonwealth/client/styles/components/component_kit/new_designs/cw_text_input.scss]: > Apparently changing it to OldTextInput is not enough. Some styles are still messed up here and there, because your inputs are not 100% in isolation.
> 
> 1. edit profile page:
> 
> * vertical margin are messed up
> * there is error but green text on validation text
> * validation text is misaligned
> 
> image
> 
> 2. create thread page:
> 
> * both inputs are full width in block styles
> 
> image
> 
> In both cases your changes are affecting the old styles which is not desired. Probably there are more places like this in the app where inputs are messed up

1. edit profile page

- When you change the page to use the new component (importing from `**/component_kit/new_designs/CWTextInput`) the 'vertical margin' you mentioned probably depends on the isCompact props, which defines its class, either `large` or `small`.
- validation text and color fixed
- validation text was misaligned on old component. Working fine on new one.

image

2. create thread page

Same here: changing the component to new CWTextInput + using isCompact and fullWidth props make the two inputs appear as intended:

image

image

**Comment by martins87** [comment]
> Component itself looks pretty ready but new input affects the old input styles so there are places in the app that are broken in styles which I guess is not desirable

Thanks again for the PR review.
The places on the app that use the new TextInput must use `isCompact` and/or `fullWidth` props to have behaviour as before.
These two were introduced as requested by design:
- `isCompact`: small and large
- `fullWidth` to make the input have the width set by outer div

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTextInput.tsx
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTextInput.tsx:66]: break this file into multiple files so the structure look like this

components/component_kit/new_designs/CWTextInput => directory, and inside:
- CWTextInput.tsx => file component
- CWTextInput.scss => style file 
- MessageRow.tsx => file component
- MessageRow.scss => style file
- useTextInputWithValidation.ts => hook file
- index.ts => import/export of CWTextInput

**Comment by masvelio** [comment]
@martins87 One change needed with file/directory structure. Otherwise, looks good. 

@Miaplacidus please decide if we need the design signoff here or we want to merge it (it is only input implementation, without replacing input across the board). 
Also @Miaplacidus we need some FF ticket that will handle swapping old inputs with new ones.

File: packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTextInput.tsx
Comment (martins87) [reply packages/commonwealth/client/scripts/views/components/component_kit/new_designs/CWTextInput.tsx]: Addressed ([here](https://github.com/hicommonwealth/commonwealth/commit/9da3b3daf835b25e7f1c3fb22d33b6d8ef916a2e) and [here](https://github.com/hicommonwealth/commonwealth/commit/0a37f530267f738895420680922b83f035da3bb9)) and tested on
- Storybook
- /components page
- edit profile page

**Comment by martins87** [comment]
> @martins87 One change needed with file/directory structure. Otherwise, looks good.
> 
> @Miaplacidus please decide if we need the design signoff here or we want to merge it (it is only input implementation, without replacing input across the board). Also @Miaplacidus we need some FF ticket that will handle swapping old inputs with new ones.

@Miaplacidus the thing about swapping old inputs with new ones is that it's not just replacing the imports on the ~40 files CWTextInput is being used.
As per design the new component introduced props `isCompact` and `fullWidth` that handle small or large size and set the component width based on outer div, respectively.
So for every case it needs to be tested.

Cc @jnaviask @MuonShot

**Review by masvelio — APPROVED** [review]

**Comment by Miaplacidus** [comment]
@masvelio @martins87 Design requests have been addressed and aren't extensive, hence additional design review isn't necessary (in general, though, it's preferred to confirm with design that your fixes address their concerns). I'll create a ticket around implementation indicating some of the issues you've brought up here. I'll tag you on it @martins87 so you can add any necessary context.

**Comment by Miaplacidus** [comment]
@martins87 @masvelio @jnaviask @MuonShot 
New card created here #4904

**Comment by masvelio** [comment]
considering above conversation, I think this PR is g2g ? @jnaviask

**Comment by jnaviask** [comment]
@masvelio I want to confirm -- are the existing styles affected? I'm worried about the edits to the `mixins/inputs.scss` file causing undesired changes. If not, then this PR is ready to go.

**Comment by martins87** [comment]
> @masvelio @martins87 I want to confirm -- are the existing styles affected? I'm worried about the edits to the `mixins/inputs.scss` file causing undesired changes. If not, then this PR is ready to go.

The [changes](https://github.com/hicommonwealth/commonwealth/pull/4293/files) on `client/styles/mixins/inputs.scss` were made on the `inputStyles` mixin, which is included in the following files:

image

Below are results and analysis of these changes.

## 1. CWTextInput.scss

Changes for the current ticket. Have been approved. Can be seen on `/components`

## 2. cw_select_list.scss

The SelectList component (`client/scripts/views/components/component_kit/cw_select_list.tsx`) [includes](https://github.com/hicommonwealth/commonwealth/blob/master/packages/commonwealth/client/styles/components/component_kit/cw_select_list.scss) on its `.scss` file the `inputStyles` and here are the result of all new styles applied:

image

image

image

## 3. cw_text_area.scss

Also [includes](https://github.com/hicommonwealth/commonwealth/blob/master/packages/commonwealth/client/styles/components/component_kit/cw_text_area.scss#L16) `inputStyle` mixin.

According to [Figma designs](https://www.figma.com/file/eIVp33a1oCu0AtcLwSbGjr/%F0%9F%9A%A7-Components-and-Patterns?type=design&node-id=339%3A4862&t=hitJ0sAOSoNBXjbx-1), the text area input field should have all new styling applied to all statuses (i.e. Default, Disabled, Hover, Focus, Filled, isInvalid). Here are the styles once the component uses the new `inputStyles` mixin styling:

### Default

image

### Disabled

image

### Hover

image

### Focus

image

image

### Filled

image

### isInvalid

!!! **NOT TESTED** !!!

## 4. cw_text_input.scss

Won't be used anymore.

---

## CWTextInput use on other components

The `CWTextInput` is used in some components (I didn't check them all yet), like `CWDropdown`.
Here is the component with old and new `CWTextInput`:

#### CWDropdown old style:

image

#### CWDropdown with new CWTextInput (width to be adjusted):

image

We would have to check the effects of new CWTextInput on components that use it.

---

Cc @jnaviask @MuonShot @masvelio @Miaplacidus @jessmart1213

**Comment by masvelio** [comment]
Thanks for the explanation @martins87. Apparently there are the changes of the current styles @jnaviask, but if they look good ([1] align with new designs and [2] do not break layout) I think we can get this in and close this 2,5 months-lasting PR

**Comment by jnaviask** [comment]
@masvelio @martins87 Thanks for the heads up. Will give this PR a review today, let's get it in the next release.

**Review by jnaviask — APPROVED** [review]

**Comment by jessmart1213** [comment]
Hi @martins87 Where can I test this? Asking because it's not tied to a specific page in demo, so where can I see all these states?

**Comment by Miaplacidus** [comment]
@jessmart1213 @martins87 
That should be available on Chromatic!

**Comment by martins87** [comment]
> Hi @martins87 Where can I test this? Asking because it's not tied to a specific page in demo, so where can I see all these states?

@jessmart1213 it's on production, you can test the TextInput component on [/components](https://commonwealth.im/components)

**Comment by jessmart1213** [comment]
### Design Review
Hi @martins87, I translated my feedback from the notion doc onto here. 

For transparency, when reviewing this work I noticed various instances where the work did not match the designs and when I went to look at the designs, I found the designs themselves were unclear. So I clarified some states and created a loom to to show the updates. 

---

## Curious questions

- What is the reasoning behind having default and compact inputs at custom widths 250px, 275px, and 300px?
    - I don’t see this reflected in the designs, so wondering if there was Eng purpose for this?
    - Also all components, including inputs/form fields should have lengths divisible by 4, unless adjusting for responsive behavior.
    

## For both compact and default sizes

- Update ‘field label’ font weight to Caption/Medium.
- Ensure disabled background color is Neutrals / 50 or #F7F7F7 not Neutrals / 100.
- Icon color, Update blue icons to Neutrals / 400.
image
- Update spacing between error/success icons and message to 4px. Currently it is 8px.
image
image


## Error management (also for both form sizes)

image


- Update so that the Invalid state is triggered immediately once user enters an invalid character. Currently Invalid state is triggered when user clicks away from the form field.

image


- Update to match ‘Form field’ state `Typing - Invalid` which can be seen [[here](https://www.figma.com/file/eIVp33a1oCu0AtcLwSbGjr/%F0%9F%9A%A7-Components-and-Patterns?type=design&node-id=337%3A3535&mode=design&t=3NdJagkEObnyAZh1-1)](https://www.figma.com/file/eIVp33a1oCu0AtcLwSbGjr/%F0%9F%9A%A7-Components-and-Patterns?type=design&node-id=337%3A3535&mode=design&t=3NdJagkEObnyAZh1-1). I updated these design components as it was previously unclear how the form fields should behave when an invalid input is recognized as a user types as opposed to when a user has clicked away from the input.

image


- Update to match ‘Form field’ state `Typing - Valid` which can be seen [[here](https://www.figma.com/file/eIVp33a1oCu0AtcLwSbGjr/%F0%9F%9A%A7-Components-and-Patterns?type=design&node-id=337%3A3535&mode=design&t=3NdJagkEObnyAZh1-1)](https://www.figma.com/file/eIVp33a1oCu0AtcLwSbGjr/%F0%9F%9A%A7-Components-and-Patterns?type=design&node-id=337%3A3535&mode=design&t=3NdJagkEObnyAZh1-1).

##Loom prototype walkthrough for error management
https://www.loom.com/share/7d9b754cdb804f6d9506a633b8b9d3aa?sid=b89a79da-c88e-4150-83c4-f38b89c72959
I created a loom for expected error management behavior. Please watch! 

---

Please ask me questions if anything is unclear!

**Pull request merged**
